monitor/utils/events.js

```diff
@@ -96,11 +97,40 @@ async function main(mode) {
       }
     })).map(normalizeEvent)
 
+    const foreignBridgeSwapped = new web3Foreign.eth.Contract([TOKENS_SWAPPED_EVENT_ABI], COMMON_FOREIGN_BRIDGE_ADDRESS)
+    const tokensSwappedEvents = (await getPastEvents(foreignBridgeSwapped, {
+      event: 'TokensSwapped',
+      fromBlock: MONITOR_FOREIGN_START_BLOCK,
+      toBlock: foreignBlockNumber
+    })).map(normalizeEvent)
+
+    // Get transfer events for each previous erc20
+    await Promise.all(
+      tokensSwappedEvents.map(async swap => {
+        const previousERC20 = new web3Foreign.eth.Contract(ERC20_ABI, swap.recipient)
+
+        const previousTransferEvents = (await getPastEvents(previousERC20, {
+          event: 'Transfer',
```

Comment:
What if some `Transfer` appears as result of a token swap. E.g. `<some first token>` -> `some another token` -> `some third token`? Seems that:
  - such `Transfer` should no be taken into account
  - a chain of swap operations should be investigated recursively

What do you think?

Reply:
I'm not sure I fully understood the scenario.
In the code we get all the `TokensSwapped` events, then for each token address that was swapped we get all the transfers to the bridge here:
https://github.com/poanetwork/tokenbridge/blob/a52b23cc0f4aa7e58bda9cf034d9059cf2192c67/monitor/utils/events.js#L112-L124

Then, once we have all the transfers from previous and current tokens addresses, we filter them to exclude those that were emitted as a result of a swap:
https://github.com/poanetwork/tokenbridge/blob/a52b23cc0f4aa7e58bda9cf034d9059cf2192c67/monitor/utils/events.js#L131-L137

Reply:
Yes, you are right. I have missed the fact that you get `TokensSwapped` events for `COMMON_FOREIGN_BRIDGE_ADDRESS`.

So, now I see that the following sequence of the events:
```
B.Transfer(0, bridge, <some value>
bridge.TokensSwapped(A, B, <some value>)
C.Transfer(0, bridge, <some value>
bridge.TokensSwapped(B, C, <some value>)
D.Transfer(0, bridge, <some value>
bridge.TokensSwapped(C, D, <some value>)
```
will be handled correctly.